### ISSUE TITLE
Websocket Close reason

### ISSUE THREAD

**Issue by asgoel** [django/daphne#374]
Are there plans to add support for the close reason as well as the close code for websockets? Have a use case for it, and was just curious.

**Comment by carltongibson** [comment]
This would be a topic to raise as a change to the ASGI spec here no? 

https://asgi.readthedocs.io/en/latest/specs/www.html#close-send-event

**Comment by asgoel** [comment]
I'm not sure? It looks like uvicorn does support it: https://github.com/encode/uvicorn/blob/c87c7834b77583efd5963dfddb08681206701d33/uvicorn/protocols/websockets/websockets_impl.py#L266-L270

**Comment by soerenbe** [comment]
I  ran into the same issue. This is implemented in the `autobahn` library and not in `daphne`. I have created an issue there. See https://github.com/crossbario/autobahn-python/issues/1642

**Comment by rudosch** [comment]
Are there any news about this issue?
It is a pity that channels allows you to specify a reason string, but daphne does not pass it on.

**Comment by alexanderjulmer** [comment]
I agree that sending custom codes could be useful. In my case they could indicate missing authentication or authorization. Of course this can also be achieved by accepting the connection and then immediately closing it with a custom error code. I guess  one could argue that such information goes beyond establishing a socket and should be sent via an established socket. 

Nonetheless it seems a little unintuitive. I wonder whether accepting and closing immediately clashes with any architecture decision of `channels` or whether it is an acceptable approach. If it isn't, I am happy to suggest a change to the ASGI spec. 

@soerenbe could you elaborate why this behaviour results from `autobahn` rather than `daphne`? As far as I can tell, it's implemented here when `serverReject` doesn't accept any code.

https://github.com/django/daphne/blob/8b5696768f1fcbe5e17d69a8c294949c2897c711/daphne/ws_protocol.py#L191-L195
